### PR TITLE
[torchbind] Improve IValue custom class API and remove most Capsule stuff

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34933 [torchbind] Enable tests on rocm
* #34906 [torchbind] Add more comprehensive docscrings
* **#34848 [torchbind] Improve IValue custom class API and remove most Capsule stuff**

This makes it so that `IValue(intrusive_ptr<T>)` where `T` inherits from `CustomClassHolder` constructs the proper Object with capsule attribute. It removes the behavior where that constructs a Capsule directly and moves that to `IValue::make_capsule` where it's less of a footgun. Also introduces `toCustomClass<T>()` which will do the inverse, it'll take the full `Object` and pull the custom class out from the capsule, returning the `intrusive_ptr`

Differential Revision: [D20480514](https://our.internmc.facebook.com/intern/diff/D20480514)

### CODE REVIEW

File: aten/src/ATen/core/ivalue.h [hunk 1]
```diff
@@ -207,14 +207,25 @@ struct CAFFE2_API IValue final {
...
 
+  // Custom C++ classes
+  template <typename T, std::enable_if_t<std::is_base_of<torch::CustomClassHolder, T>::value, int> = 0>
+  IValue(intrusive_ptr<T> custom_class);
+  bool isCustomClass() const;
+  template <typename T>
+  c10::intrusive_ptr<T> toCustomClass() &&;
+  template <typename T>
+  c10::intrusive_ptr<T> toCustomClass() const &;
+
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.h]: Normally when we add a specific `toX` we also add to the templated `.to<X>`. Does `.to<MyCustomClass>()` work?

File: aten/src/ATen/core/ivalue_inl.h [hunk 2]
```diff
@@ -521,14 +522,40 @@ std::vector<Elem> generic_to(
...
+template <typename T>
+c10::intrusive_ptr<T> IValue::toCustomClass() && {
+  static_assert(std::is_base_of<torch::CustomClassHolder, T>::value == true,
+    "toCustomClass requires that template parameter T must inherit "
+    "from torch::CustomClassHolder");
+  auto obj = toObject();
+  TORCH_CHECK(obj->slots().size() == 1,
+              "Tried to cast IValue to custom class but it did "
+              "not contain a custom class!");
+  // FIXME: this cast isn't actually checked!
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue_inl.h]: Options:
1. fix before land
2. name this unsafeToCustomClass until it checks this. Our users will find this method and try to call it.

File: aten/src/ATen/core/ivalue_inl.h [hunk 3]
```diff
@@ -774,10 +801,32 @@ inline IValue::IValue(c10::intrusive_ptr<ivalue::PyObjectHolder> v)
...
+  if (!c10::isCustomClassRegistered<c10::intrusive_ptr<T>>()) {
+    throw c10::Error(
+        "Trying to instantiate a class that isn't a registered custom class.",
+        "");
+  }
+  auto classType = c10::getCustomClassType<c10::intrusive_ptr<T>>();
+  auto ivalue_obj = c10::ivalue::Object::create(
+      c10::StrongTypePtr(nullptr, classType), /*num_slots=*/1);
+  ivalue_obj->setAttr(
+      "capsule",
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue_inl.h]: setSlot would be faster

File: torch/custom_class.h [hunk 4]
```diff
@@ -74,10 +74,8 @@ class class_ {
                                                // torch::init<...>()
     auto func = [](c10::tagged_capsule<CurClass> self, Types... args) {
       auto classObj = c10::make_intrusive<CurClass>(args...);
-      auto genericPtr = c10::static_intrusive_pointer_cast<torch::CustomClassHolder>(std::move(classObj));
-      auto capsule = c10::IValue(std::move(genericPtr));
-      auto object = std::move(self.ivalue).toObject();
-      object->setSlot(0, std::move(capsule));
+      auto object = self.ivalue.toObject();
```
Comment (zdevito) [inline torch/custom_class.h]: nit: std::move was dropped here.

**Review by zdevito — APPROVED** [review]
Looks good, thanks!

File: aten/src/ATen/core/ivalue.h (hunk 1 above)
Comment (jamesr66a) [reply aten/src/ATen/core/ivalue.h]: it ends up being `.to<c10::intrusive_ptr<MyCustomClass>>()`. Maybe to normalize it we should kill the need to put `intrusive_ptr` there, I can look into it

File: torch/custom_class.h (hunk 4 above)
Comment (jamesr66a) [reply torch/custom_class.h]: I don't think `std::move()` was correct here, it was added as a codemod and was moving this IValue out, when the `__init__` method should just modify it in place

**Pull request closed without merging**